models/base/document.py

```diff
@@ -175,20 +275,85 @@ def reload(cls) -> None:
+                    # Load all indices referring to this document to relink referrers
+                    cls.__restore_referrer(cls.__objects[key])
         except FileNotFoundError:
             cls.__objects = {}
 
-    def save(self) -> None:
+    @classmethod
+    def __restore_reference(cls, value):
+        """
+        Restore deferred references to other IndexedDocuments recursively.
+        :param value: a DeferredReference, Document, ReferenceSet, or field value
+        :return: the restored value, or None if it was a field value
+        """
+        if isinstance(value, cls.DeferredReference):
+            return value.restore()
+        elif isinstance(value, Document):
+            for field_name, field in value._fields.items():
+                restored_field = cls.__restore_reference(getattr(value, field_name))
+                if restored_field is not None:
+                    setattr(value, field_name, restored_field)
+            return value
+        elif isinstance(value, ReferenceSet):
+            return [cls.__restore_reference(doc) for doc in value]
+        else:
+            return None
+
+    @classmethod
+    def __restore_referrer(cls, obj) -> None:
+        """
+        Load all indices referring to this document to relink referrers.
+        :param obj: an unpickled IndexedDocument instance
+        """
+        if getattr(obj, '_referrer_roots', None):
+            for mod, classname in obj._referrer_roots:
+                __import__(mod)
+                index = pickle._getattribute(sys.modules[mod], classname)[0]
+                index.check_and_load_data()
+            delattr(obj, '_referrer_roots')
+
+    @classmethod
+    def check_and_load_data(cls):
+        """
+        Check if the index is loaded, if not, load it.
+        """
+        if cls.__objects is None:
+            cls.reload()
+
+    def _persist(self) -> None:
         """
         Save all documents of the same type (i.e. the index) to disk.
-        Also save all root-level documents referencing this document.
+        If the class is a subclass of IndexedDocument, also persist to parent indices.
         """
-        self.__class__.__objects[self.key] = self
+        self.__class__.check_and_load_data()
         os.makedirs(os.path.dirname(self._persistence_path), exist_ok=True)
         with open(self._persistence_path, 'wb') as f:
-            pickle.dump(self.__class__.__objects, f)
-        super().save()
+            self.Pickler(f, self.__class__).dump(self.__class__.__objects)
+
+        # Also save to parent indices
+        for persistence_base in self._persistence_bases:
+            persistence_base.__index_subclass(self)
+
+    def _get_root_document(self):
+        return self
+
+    @classmethod
+    def __index_subclass(cls, child: Document) -> None:
+        """
+        Save a child document to the parent index.
+        to be called from subclass.
```

Comment:
A specific example in the context of this system may be useful to help guide the reader.

Reply:
Comment updated. 
I have not included an example because this private method is only used in one place in the same class.